### PR TITLE
Use Best Loss

### PR DESCRIPTION
**Pull request by jq**
@deerzq @yolken

### CODE REVIEW

File: training/src/main/scala/com/airbnb/aerosolve/training/AdditiveModelTrainer.scala [hunk 1]
```diff
@@ -73,13 +81,25 @@ object AdditiveModelTrainer {
...
+        sc.accumulator(0),
+        sc.accumulator(0))
       val modelBC = sc.broadcast(model)
-      model = sgdTrain(transformed, paramsBC, modelBC)
+      model = sgdTrain(transformed, sgdParams, modelBC)
       modelBC.unpersist()
-
-      TrainingUtils.saveModel(model, output)
+      val newLoss = sgdParams.loss.value/sgdParams.exampleCount.value
+      if (params.loss.useBestLoss && newLoss < loss) {
```
Comment (yinlou) [inline training/src/main/scala/com/airbnb/aerosolve/training/AdditiveModelTrainer.scala]: Typically using the relevance metric (like the loss here) is not a good idea (relevance loss is an approximate for business metric and best relevance loss does not usually give you the best business metric value). Directly using business metric is a better alternative.

Comment (jq) [inline training/src/main/scala/com/airbnb/aerosolve/training/AdditiveModelTrainer.scala]: agree, so I put a useBestLoss here, it is optional.

Comment (yinlou) [inline training/src/main/scala/com/airbnb/aerosolve/training/AdditiveModelTrainer.scala]: Great!

File: training/src/main/scala/com/airbnb/aerosolve/training/AdditiveModelTrainer.scala [hunk 2]
```diff
@@ -32,11 +32,20 @@ import scala.util.Try
 object AdditiveModelTrainer {
   private final val log: Logger = LoggerFactory.getLogger("AdditiveModelTrainer")
 
+  case class SgdParams(paramsBC: Broadcast[AdditiveTrainerParams],
+                      exampleCount: Accumulator[Long],
```
Comment (deerzq) [inline training/src/main/scala/com/airbnb/aerosolve/training/AdditiveModelTrainer.scala]: indent

File: training/src/main/scala/com/airbnb/aerosolve/training/AdditiveModelTrainer.scala [hunk 3]
```diff
@@ -73,13 +81,30 @@ object AdditiveModelTrainer {
...
+        paramsBC,
+        sc.accumulator(0),
+        sc.accumulator(0))
       val modelBC = sc.broadcast(model)
-      model = sgdTrain(transformed, paramsBC, modelBC)
+      model = sgdTrain(transformed, sgdParams, modelBC)
       modelBC.unpersist()
-
-      TrainingUtils.saveModel(model, output)
+      val newLoss = sgdParams.loss.value/sgdParams.exampleCount.value
```
Comment (deerzq) [inline training/src/main/scala/com/airbnb/aerosolve/training/AdditiveModelTrainer.scala]: space

File: training/src/main/scala/com/airbnb/aerosolve/training/AdditiveModelTrainer.scala [hunk 4]
```diff
@@ -73,13 +81,30 @@ object AdditiveModelTrainer {
...
+           loss = newLoss
+         } else {
+           log.info(s"iterations $i loss $newLoss < best lost $loss count = $sgdParams.exampleCount.value")
+         }
+      } else {
+        TrainingUtils.saveModel(model, output)
+        log.info(s"iterations $i Loss = $newLoss count = $sgdParams.exampleCount.value")
+      }
     }
     model
```
Comment (deerzq) [inline training/src/main/scala/com/airbnb/aerosolve/training/AdditiveModelTrainer.scala]: you are still returning the model in the last iteration, then the saveModel in `trainAndSaveToFileEarlySample` would overwrite whatever model you have saved with the last model

File: training/src/main/scala/com/airbnb/aerosolve/training/AdditiveModelTrainer.scala [hunk 5]
```diff
@@ -210,6 +238,8 @@ object AdditiveModelTrainer {
           output.put((family._1, feature._1), feature._2)
         })
       })
+    sgdParams.exampleCount.+=(lossCount)
+    sgdParams.loss.+=(lossTotal)
```
Comment (deerzq) [inline training/src/main/scala/com/airbnb/aerosolve/training/AdditiveModelTrainer.scala]: could you write this as `sgdParams.exampleCount += lossCount`?

**Comment by jq** [comment]
@deerzq thanks for review, all updated

**Comment by deerzq** [comment]
LGTM
